### PR TITLE
GH-37591: [MATLAB] Make `arrow.type.Type` inherit from `matlab.mixin.Heterogeneous`

### LINKED ISSUE #37591: [MATLAB] Make `arrow.type.Type` inherit from `matlab.mixin.Heterogeneous`
### Describe the enhancement requested

We should modify `arrow.type.Type` to inherit from `matlab.mixin.Heterogeneous` so that it's possible to create a nonscalar array containing different concrete subclasses of `arrow.type.Type`. 

**Example**
```matlab
>> float32Type = arrow.float32();
>> timestampType = arrow.timestamp(TimeZone="Pacific/Fiji");
>> typeArray = [float32Type timestampType]

typeArray = 

  1×2 heterogeneous FixedWidthType (Float32Type, TimestampType) array with properties:

    ID
```

Currently, running the example above results in an error:

```matlab
>> float32Type = arrow.float32();
>> timestampType = arrow.timestamp(TimeZone="Pacific/Fiji");
>> typeArray = [float32Type timestampType]
Error using horzcat
The following error occurred converting from arrow.type.TimestampType to arrow.type.Float32Type:
Invalid argument at position 1. Value must be of type libmexclass.proxy.Proxy or be convertible to libmexclass.proxy.Proxy.
```

### Component(s)

MATLAB

### PR DESCRIPTION
**Pull request by sgilmore10**
### Rationale for this change

We should modify `arrow.type.Type` to inherit from `matlab.mixin.Heterogeneous` so that it's possible to create a nonscalar array containing different concrete subclasses of `arrow.type.Type`. 

**Example**
```matlab
>> float32Type = arrow.float32();
>> timestampType = arrow.timestamp(TimeZone="Pacific/Fiji");
>> typeArray = [float32Type timestampType]

typeArray = 

  1×2 heterogeneous FixedWidthType (Float32Type, TimestampType) array with properties:

    ID
```

Currently, running the example above results in an error:

```matlab
>> float32Type = arrow.float32();
>> timestampType = arrow.timestamp(TimeZone="Pacific/Fiji");
>> typeArray = [float32Type timestampType]
Error using horzcat
The following error occurred converting from arrow.type.TimestampType to arrow.type.Float32Type:
Invalid argument at position 1. Value must be of type libmexclass.proxy.Proxy or be convertible to libmexclass.proxy.Proxy.
```

### What changes are included in this PR?

1. Updated `arrow.type.Type` to inherit from `matlab.mixin.Heterogeneous`. Previously, it only inherited from `matlab.mixin.CustomDisplay`.
2. Because `arrow.type.Type` is now `heterogeneous`, its subclasses cannot override the protected method `getPropertyGroups` - which is inherited from `matlab.mixin.CustomDisplay`. This method must be `Sealed` and defined on `Type` because only `Sealed` methods can be called on nonscalar `heterogeneous` arrays. To get around this limitation, I defined a new abstract template method called `getDisplayPropertyGroups`, which `getPropertyGroups` invokes if the array is either scalar or nonscalar and `homogeneous`. Refer to [this link](https://www.mathworks.com/help/matlab/matlab_oop/custom-display-for-heterogeneous-arrays.html) for more details. 

**Displaying a nonscalar *heterogeneous* `arrow.type.Type`**

```matlab
>> float32Type = arrow.timestamp(TimeUnit="Second");
>> timestampType = arrow.timestamp(TimeZone="Pacific/Fiji");
>> typeArray = [float32Type timestampType]

typeArray = 

  1×2 heterogeneous FixedWidthType (Float32Type, TimestampType) array with properties:

    ID
```

**Displaying a nonscalar *homogeneous* `arrow.type.Type`**

```matlab
>> timestampType1 = arrow.timestamp(TimeUnit="Second");
>> timestampType2 = arrow.timestamp(TimeZone="Pacific/Fiji");
>> typeArray = [timestampType1 timestampType2]

typeArray = 

  1×2 TimestampType array with properties:

    ID
    TimeUnit
    TimeZone
```

### Are these changes tested?

Yes

1. Added a new test class in `test/arrow/type` called `tDisplay.m` 
2. Added display unit tests for all concrete subclasses of `arrow.type.Type` if they did not already exist. 
3. Added a new test class in `test/arrow/type` called `tIsEqual.m`. This class contains unit tests verifying the `isequal` method of `arrow.type.Type` behaves as expected when called on heterogeneous arrays.

### Are there any user-facing changes?

Yes. Users can now create nonscalar  arrays with different concrete subclasses of `arrow.type.Type`.

### Future Directions

1. Move all display unit tests to `tDisplay.m`
2. Add an `arrow.type.NumericType` class from which all numeric types inherit. This class will define `getDisplayPropertyGroups` for all numeric types.

* Closes: #37591

### CODE REVIEW

File: matlab/test/arrow/type/tDisplay.m [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    if opts.BoldFont
+        link = compose("<a href=""matlab:helpPopup %s"" style=""font-weight:bold"">%s</a>", ...
+            opts.FullClassName, opts.ClassName);
+    else
+        link = compose("<a href=""matlab:helpPopup %s"">%s</a>", opts.FullClassName, opts.ClassName);
+    end
+end
+
+function sizeString = makeSizeString(arraySize)
```
Comment (kevingurney) [inline matlab/test/arrow/type/tDisplay.m:130]: Maybe we should rename this to something like `makeDimensionDisplay`?

**Review by kevingurney — CHANGES_REQUESTED** [review]
Overall, this looks good to me!

Just one high level comment - I think it would make sense to move all of the type-specific display tests you added into `tDisplay` for simplicity (this will also allow you to use your new helper method `makeDisplayLink` in these tests).

File: matlab/test/arrow/type/tDisplay.m [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            actualDisplay = evalc('disp(typeArray)');
+            testCase.verifyDisplay(actualDisplay, expectedDisplay);
+        end
+
+        function NonScalarArraySameTypes(testCase)
+            % Verify the display of a scalar, homogeneous arrow.type.Type array:
+            % 
+            % Example:
+            %
+            %  1×2 heterogeneous FixedWidthType (Float32Type, TimestampType) array with properties:
```
[the remark this answers is not in the document]
Comment (sgilmore10) [reply matlab/test/arrow/type/tDisplay.m:76]: My bad. I'll fix that!

File: matlab/test/arrow/type/tDisplay.m (hunk 1 above)
Comment (sgilmore10) [reply matlab/test/arrow/type/tDisplay.m:130]: Agreed. This name isn't great. I decided to rename `makeSizeString` to `makeDimensionString` and `makeDisplayLink` to `makeLinkString`.

**Comment by sgilmore10** [comment]
> Overall, this looks good to me!
> 
> Just one high level comment - I think it would make sense to move all of the type-specific display tests you added into `tDisplay` for simplicity (this will also allow you to use your new helper method `makeDisplayLink` in these tests).

Good idea! I just did this.

File: matlab/test/arrow/type/tTypeDisplay.m [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+    if opts.BoldFont
+        link = compose("<a href=""matlab:helpPopup %s"" style=""font-weight:bold"">%s</a>", ...
+            opts.FullClassName, opts.ClassName);
+    else
+        link = compose("<a href=""matlab:helpPopup %s"">%s</a>", opts.FullClassName, opts.ClassName);
+    end
+end
+
+function sizeString = makeDimensionString(arraySize)
```
Comment (kevingurney) [inline matlab/test/arrow/type/tTypeDisplay.m:253]: Sorry, I should have added this comment before. Could we rename the return value to `dimensionString`?

Comment (sgilmore10) [reply matlab/test/arrow/type/tTypeDisplay.m:253]: ah yeah. Will do.

**Review by kevingurney — APPROVED** [review]

**Comment by kevingurney** [comment]
+1

**Pull request merged**